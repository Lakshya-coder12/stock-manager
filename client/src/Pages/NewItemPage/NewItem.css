@media (max-width: 480px) {
  .form {
    width: 75vw;
    border-radius: 5px;
    padding: 10px;
    background-color: white;
    box-shadow: 3px 3px #babab8;
    margin: 0 auto;
    margin-top: 100px;
  }
}

@media (min-width: 480px) and (max-width: 950px) {
  .form {
    width: 50vw;
    border-radius: 5px;
    padding: 10px;
    background-color: white;
    box-shadow: 3px 3px #babab8;
    margin: 0 auto;
    margin-top: 100px;
  }
}
@media (min-width: 950px) {
  .form {
    width: 30vw;
    border-radius: 5px;
    padding: 10px;
    background-color: white;
    box-shadow: 3px 3px #babab8;
    margin: 0 auto;
    margin-top: 100px;
  }
}

.input {
  outline: none;
  margin-top: 10px;
}

.label {
  color: #1976c2;
  font-weight: 600;
  font-size: large;
}
.input-element {
  margin-top: 10px;
  display: flex;
  flex-direction: column;
}

.form-btn {
  margin-top: 20px;
}

.username {
  color: #1976c2;
  font-weight: 600;
  margin-right: 5px;
}
