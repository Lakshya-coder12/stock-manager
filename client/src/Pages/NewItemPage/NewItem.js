import React, { Component } from "react";
import "./NewItem.css";
import Button from "../../Components/Button/Button";
// import { connect } from "react-redux";
// import { insertItem } from "../../redux/actions/ItemActions";
import PropTypes from "prop-types";
import jwt_decode from "jwt-decode";

export class NewItem extends Component {
  constructor(props) {
    super(props);
    this.state = {
      name: "",
      companyName: "",
      stock: 0,
      username: "",
    };
  }

  componentDidMount() {
    const token = localStorage.getItem("jwt");
    if (!token) {
      this.props.history.push("/login");
    } else {
      const user = jwt_decode(token);
      const username = user.user.name;
      this.setState({
        username: username,
      });
    }
  }

  handleInputChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  };
  handleRedirect = () => {
    this.props.history.push("/home");
  };

  handleLogout = () => {
    this.props.logout(this.props.history);
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const formData = {
      name: this.state.name,
      companyName: this.state.companyName,
      stock: this.state.stock,
    };
    this.props.insertItem(formData, this.props.history);
  };
  render() {
    const { name, companyName, stock } = this.state;
    return (
      <>
        <Button className="abs-button" onClick={this.handleRedirect}>
          Home
        </Button>
        <div className="form">
          <div className="input-element">
            <label className="label">Name</label>
            <input
              name="name"
              onChange={this.handleInputChange}
              value={this.state.name}
              className="input"
              type="text"
            />
          </div>
          <div className="input-element">
            <label className="label">Company Name</label>
            <input
              name="companyName"
              className="input"
              onChange={this.handleInputChange}
              value={this.state.companyName}
              type="text"
            />
          </div>
          <div className="input-element">
            <label className="label">Stock</label>
            <input
              name="stock"
              onChange={this.handleInputChange}
              value={this.state.stock}
              className="input"
              type="number"
              step={1}
            />
          </div>
          <div className="form-btn">
            <Button
              onClick={
                !stock || !companyName || !name ? null : this.handleSubmit
              }
              id="btn1"
            >
              Add Item
            </Button>
          </div>
        </div>
        <div className="logout">
          <span className="username">{this.state.username}</span>
          <Button onClick={this.handleLogout}>Logout</Button>
        </div>
      </>
    );
  }
}

NewItem.propTypes = {
  insertItem: PropTypes.func.isRequired,
  logout: PropTypes.func.isRequired,
};

// const mapDispatchToProps = (dispatch) => {
//   return {
//     insertItem: (data, history) => dispatch(insertItem(data, history)),
//   };
// };

// export default connect(null, mapDispatchToProps)(NewItemPage);

export default NewItem;
