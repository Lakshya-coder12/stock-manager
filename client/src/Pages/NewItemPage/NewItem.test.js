import { configure, shallow } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import NewItem from "./NewItem";

configure({ adapter: new Adapter() });

let wrapper;

beforeAll(() => {
  const localStorageMock = (() => {
    let store = {};
    return {
      getItem(key) {
        return store[key];
      },
      setItem(key, value) {
        store[key] = value.toString();
      },
      clear() {
        store = {};
      },
      removeItem(key) {
        delete store[key];
      },
    };
  })();
  Object.defineProperty(window, "localStorage", {
    value: localStorageMock,
  });
});

afterAll(() => {
  window.localStorage.clear();
  delete window.localStorage;
});

beforeEach(() => {
  const insertItemMock = jest.fn();
  const logoutMock = jest.fn();
  const pushMock = jest.fn();
  wrapper = shallow(
    <NewItem
      insertItem={insertItemMock}
      logout={logoutMock}
      history={{ push: pushMock }}
    />
  );
});

describe("Testing New Item page", () => {
  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("check State", () => {
    expect(wrapper.state("name")).toEqual("");
  });

  it("check for name field", () => {
    wrapper.setState({ name: "Mouse" });
    expect(wrapper.state("name")).toEqual("Mouse");
  });

  it("check for companyName field", () => {
    wrapper.setState({ companyName: "Razer" });
    expect(wrapper.state("companyName")).toEqual("Razer");
  });

  it("check for stock field", () => {
    wrapper.setState({ stock: 100 });
    expect(wrapper.state("stock")).toEqual(100);
  });

  it("check for username field", () => {
    wrapper.setState({ username: "Lakshya" });
    expect(wrapper.state("username")).toEqual("Lakshya");
  });

  it("Button not works when any field is empty", () => {
    const instance = wrapper.instance();
    jest.spyOn(instance, "handleSubmit");
    wrapper.setState({ name: "Shoes", stock: 100, companyName: "" });
    wrapper.find("#btn1").simulate("click");
    expect(instance.handleSubmit).toHaveBeenCalledTimes(0);
  });

  it("Button works when every input field is valid", () => {
    const instance = wrapper.instance();
    jest.spyOn(instance, "handleSubmit");
    wrapper.setState({ name: "Shoes", stock: 100, companyName: "Nike" });
    wrapper.find("#btn1").simulate("click", {
      preventDefault: () => {},
    });
    expect(instance.handleSubmit).toHaveBeenCalledTimes(1);
  });

  it("Checks for token in localStorage", () => {
    window.localStorage.setItem(
      "jwt",
      "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c"
    );
    expect(window.localStorage.getItem("jwt")).toEqual(
      "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.SflKxwRJSMeKKF2QT4fwpMeJf36POk6yJV_adQssw5c"
    );
  });
});
