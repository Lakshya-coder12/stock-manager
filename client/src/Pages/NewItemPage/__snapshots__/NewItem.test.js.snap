// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing New Item page matches snapshot 1`] = `
<Fragment>
  <Button
    className="abs-button"
    onClick={[Function]}
  >
    Home
  </Button>
  <div
    className="form"
  >
    <div
      className="input-element"
    >
      <label
        className="label"
      >
        Name
      </label>
      <input
        className="input"
        name="name"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="input-element"
    >
      <label
        className="label"
      >
        Company Name
      </label>
      <input
        className="input"
        name="companyName"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="input-element"
    >
      <label
        className="label"
      >
        Stock
      </label>
      <input
        className="input"
        name="stock"
        onChange={[Function]}
        step={1}
        type="number"
        value={0}
      />
    </div>
    <div
      className="form-btn"
    >
      <Button
        id="btn1"
        onClick={null}
      >
        Add Item
      </Button>
    </div>
  </div>
  <div
    className="logout"
  >
    <span
      className="username"
    />
    <Button
      onClick={[Function]}
    >
      Logout
    </Button>
  </div>
</Fragment>
`;
